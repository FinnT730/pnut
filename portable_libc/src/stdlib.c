#include "include/stdlib.h"

#define HEAP_SIZE 1000000

char _heap[HEAP_SIZE];
int _heap_alloc = 0;

void *malloc(size_t size) {
  char *result = _heap + _heap_alloc;
  if (_heap_alloc + size > HEAP_SIZE) return 0;
  _heap_alloc += size;
  return result;
}

void free(void *ptr) {
}

void *realloc(void *ptr, size_t size) {
  return 0; /*TODO*/
}

double strtod(const char *str, char **endptr) {
  return 0.0; /*TODO*/
}

float strtof(const char *str, char **endptr) {
  return 0.0; /*TODO*/
}

long double strtold(const char *str, char **endptr) {
  return 0.0; /*TODO*/
}

long int strtol(const char *str, char **endptr, int base) {

  char *probe = str;
  int neg;
  int n = 0;
  int nb_digits = 0;
  int d;

  while (*probe && *probe <= ' ') probe++; /* skip spaces */

  neg = *probe == '-';

  if (neg || *probe == '+') probe++;

  if (*probe == '0') {
    probe++;
    nb_digits = 1;
    if (base == 0 || base == 16) {
      if (*probe == 'x') {
        base = 16;
        probe++;
        nb_digits = 0;
      } else if (base == 0) {
        base = 8;
      }
    }
  }

  while (1) {
    d = *probe;
    if (d >= '0' && d <= '9') {
      d = d-'0';
    } else if (d >= 'A' && d <= 'Z') {
      d = d-'A'+10;
    } else if (d >= 'a' && d <= 'z') {
      d = d-'a'+10;
    } else {
      d = 99;
    }
    if (d >= base) break;
    n = n*base - d;
    nb_digits++;
  }

  if (endptr) {
    if (nb_digits == 0) {
      *endptr = str;
    } else {
      *endptr = probe;
    }
  }

  return neg ? n : -n;
}

int atoi(const char *str) {
  return strtol(str, 0, 10);
}
