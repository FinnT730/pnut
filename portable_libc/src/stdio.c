#include "include/stdio.h"

#include "include/stdlib.h"
#include "include/unistd.h"

FILE _stdin_FILE;
FILE _stdout_FILE;
FILE _stderr_FILE;
FILE _string_FILE;

FILE *stdin = &_stdin_FILE;
FILE *stdout = &_stdout_FILE;
FILE *stderr = &_stderr_FILE;

int _get_fd(FILE *f) {
  if (f == &_stdin_FILE) {
    return 0;
  } else if (f == &_stdout_FILE) {
    return 1;
  } else if (f == &_stderr_FILE) {
    return 2;
  } else {
    return f->fd;
  }
}

FILE *fopen(const char *pathname, const char *mode) {
  return 0; /*TODO*/
}

FILE *fdopen(int fd, const char *mode) {
  FILE *result = malloc(sizeof(FILE));
  if (result) {
    result->fd = fd;
  }
  return result;
}

int fclose(FILE *stream) {
  return 0; /*TODO*/
}

int fputc(int c, FILE *stream) {
  int fd = _get_fd(stream);
  if (fd == -1) {
    if (stream->string_output_len+1 < stream->string_output_buf_size) {
      stream->string_output_buf[stream->string_output_len] = c;
      ++stream->string_output_len;
    }
  } else {
    stream->buf[0] = c;
    write(fd, stream->buf, 1);
  }
  return c;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  const char *p = ptr;
  size_t n = size * nmemb;
  while (n) {
    fputc(*p, stream);
    ++p;
    --n;
  }
  return nmemb;
}

int _fputstr(const char *str, FILE *stream) {
  int result = 0;
  while (*str) {
    fputc(*str, stream);
    ++str;
    ++result;
  }
  return result;
}

int fputs(const char *s, FILE *stream) {
  _fputstr(s, stream);
  fputc('\n', stream);
  return 0;
}

int puts(const char *s) {
  return fputs(s, stdout);
}

#define SIZEOF_NUM_BUF 100
char _num_buf[SIZEOF_NUM_BUF];
char *_digits = "0123456789abcdef";

char *_int_to_str(int n, int base, int width, int force_0, int force_plus) {

  char *out = _num_buf + SIZEOF_NUM_BUF;
  int neg = n < 0;
  int has_sign = neg || force_plus;

  if (!neg) n = -n;

  *--out = 0;

  while (n <= -base) {
    *--out = _digits[-(n % base)];
    n /= base;
  }

  *--out = _digits[-(n % base)];

  if (width >= SIZEOF_NUM_BUF) width = SIZEOF_NUM_BUF-1;

  if (force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width - has_sign) {
      *--out = force_0 ? '0' : ' ';
    }
  }

  if (has_sign) *--out = neg ? '-' : '+';

  if (!force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width) {
      *--out = ' ';
    }
  }

  return out;
}

int vfprintf(FILE *stream, const char *format, va_list ap) {

  char c;
  int result = 0;
  int base;
  int width;
  int force_0;

  while ((c = *format++)) {
    if (c == '%' && (c = *format++) != '%') {
      width = 0;
      force_0 = 0;
      if (c == '0') {
        force_0 = 1;
        c = *format++;
      }
      while (c >= '0' && c <= '9') {
        width = width*10 + (c - '0');
        c = *format++;
      }
      while (c == 'l' || c == 'L') {
        c = *format++;
      }
      switch (c) {
      case 'c':
        fputc(va_arg(ap, int), stream);
        result += 1;
        break;
      case 'd':
      case 'u':
      case 'o':
      case 'x':
        base = c == 'x' ? 16 : c == 'o' ? 8 : 10;
        result += _fputstr(_int_to_str(va_arg(ap, int), base, width, force_0, 0), stream);
        break;
      case 's':
        result += _fputstr(va_arg(ap, char*), stream);
        break;
      default:
        fputc('?', stream);
        fputc('?', stream);
        fputc('?', stream);
        result += 3;
      }
    } else {
      fputc(c, stream);
      result += 1;
    }
  }

  return result;
}

int fprintf(FILE *stream, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start (ap, format);
  result = vfprintf(stream, format, ap);
  va_end(ap);

  return result;
}

int printf(const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start (ap, format);
  result = vfprintf(stdout, format, ap);
  va_end(ap);

  return result;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {

  int result;

  _string_FILE.fd = -1;
  _string_FILE.string_output_buf = str;
  _string_FILE.string_output_buf_size = size;
  _string_FILE.string_output_len = 0;
  result = vfprintf(&_string_FILE, format, ap);
  _string_FILE.string_output_buf[_string_FILE.string_output_len] = 0; /* null terminate string */

  return result;
}

int snprintf(char *str, size_t size, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start (ap, format);
  result = vsnprintf(str, size, format, ap);
  va_end(ap);

  return result;
}

int sprintf(char *str, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start (ap, format);
  result = vsnprintf(str, 999999999, format, ap);
  va_end(ap);

  return result;
}
