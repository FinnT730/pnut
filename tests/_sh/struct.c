#include <stdio.h>
#include <stdlib.h>

typedef enum Direction {
  Up,
  Down,
  Left,
  Right,
} Direction;

typedef enum {
  North,
  South,
  East,
  West,
} CardinalDirection;

// Typedef of named structure
typedef struct Rectangle {
  int w;
  int h;
} Rectangle;

// Typedef of anonymous structure
typedef struct {
  int x;
  int y;
} Point;

// Named structure declaration
struct Shape {
  Point *origin;
  Rectangle *r;
};

void f(enum Direction dir, Direction dir2) {
  printf("Direction: %d %d\n", dir, dir2);
}

void pass_as_ref(Point *pt) {
  printf("pass_as_ref: Point: %d %d\n", pt->x, pt->y);
  pt->x = 123;
  pt->y = 456;
  printf("pass_as_ref: Point: %d %d\n", pt->x, pt->y);
}

void pass_as_ref_int(int *x) {
  printf("pass_as_ref_int: %d\n", *x);
  *x = 42;
  printf("pass_as_ref_int: %d\n", *x);
}

// Show that we can have a struct pointer
struct Point *pts;

void main() {
  Direction up = Up;
  const enum Direction down = Down;
  Direction left = Left, right = Right;

  // Rectangle can be with and without the struct keyword
  struct Rectangle *r1 = malloc(sizeof(struct Rectangle));
  Rectangle *r2 = malloc(sizeof(Rectangle));
  Point *pt = malloc(sizeof(Point));
  struct Shape *shape = malloc(sizeof(struct Shape));
  struct Shape **shapes = (struct Shape **) malloc(3 * sizeof(struct Shape));
  int *p = malloc(2 * sizeof(int));
  int *arr = malloc(24);

  int i;

  // Initialize shapes array
  for (i = 0; i < 3; i++) {
    shapes[i] = malloc(sizeof(struct Shape));
    shapes[i]->origin = malloc(sizeof(Point));
    shapes[i]->r = malloc(sizeof(Rectangle));
  }

  // Assign values
  for (i = 0; i < 3; i++) {
    shapes[i]->origin->x = i;
    shapes[i]->origin->y = i * i;
    shapes[i]->r->w = i;
    shapes[i]->r->h = i * i;
  }

  // Print values
  for (i = 0; i < 3; i++) {
    printf("Shape %d: %d %d %d %d\n", i, shapes[i]->origin->x, shapes[i]->origin->y, shapes[i]->r->w, shapes[i]->r->h);
  }

  pt->x = 5;
  pt->y = 6;

  r1->w = 1;
  r1->h = 2;

  shape->origin = pt;
  shape->r = r1;

  printf("Rectangle 1: %d %d\n", r1->w, r1->h);
  printf("Point: %d %d\n", pt->x, pt->y);
  printf("Shape: %d %d %d %d\n", shape->origin->x, shape->origin->y, shape->r->w, shape->r->h);

  // Test passing as reference a struct ptr
  pass_as_ref(pt);
  printf("Point: %d %d\n", pt->x, pt->y);

  // Test passing as reference an integer from a struct
  pass_as_ref_int(&pt->x);
  printf("Point: %d %d\n", pt->x, pt->y);

  // Test passing as reference an integer from an array
  arr[12] = 12;
  printf("arr[12]: %d\n", arr[12]);
  pass_as_ref_int(&arr[12]);
  printf("arr[12]: %d\n", arr[12]);

  exit(0);
}
