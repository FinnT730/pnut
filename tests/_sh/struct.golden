Shape 0: 0 0 0 0
Shape 1: 1 1 1 1
Shape 2: 2 4 2 4
Rectangle 1: 1 2
Point: 5 6
Shape: 5 6 1 2
pass_as_ref: Point: 5 6
pass_as_ref: Point: 123 456
Point: 123 456
pass_as_ref_int: 123
pass_as_ref_int: 42
Point: 42 456
arr[12]: 12
pass_as_ref_int: 12
pass_as_ref_int: 42
arr[12]: 42
